conductr_cli/bndl_utils.py

```diff
@@ -75,74 +74,154 @@ def try_match(pattern, slice):
     elif b'/manifest.json\x00\x00\x00' in initial_chunk and b'/layer.tar\x00\x00\x00' in initial_chunk:
         # tar c on a docker folder (somewhat unreliable)
         return 'docker'
+    elif b'\x00' not in initial_chunk and initial_chunk != b'' and data_is_bundle_conf(initial_chunk):
+        return 'bundle'
+    elif data_is_zip(initial_chunk):
+        return 'bundle'
+    elif data_is_tar(initial_chunk):
+        # TAR marker
+        return 'bundle'
     else:
         return None
 
 
-def load_bundle_args_into_conf(config, args):
-    if args.name is not None:
-        config.put('name', args.name)
+def data_is_bundle_conf(data):
+    try:
+        ConfigFactory.parse_string(data.decode('UTF-8'))
+        return True
+    except:
+        return False
 
-    for argument, bundle_key in mappings.items():
-        value = getattr(args, bundle_key, None)
 
-        if value is not None:
-            config.put(bundle_key, value)
+def data_is_tar(data):
+    return len(data) >= 262 and data[257:262] == b'ustar'
 
-    if 'roles' not in config:
-        config.put('roles', BNDL_DEFAULT_ROLES)
 
-    if 'compatibilityVersion' not in config:
-        config.put('compatibilityVersion', BNDL_DEFAULT_COMPATIBILITY_VERSION)
+def data_is_zip(data):
+    return \
+        data.startswith(b'PK\x03\x04') or \
+        data.startswith(b'PK\x05\x06') or \
+        data.startswith(b'PK\x07\x08')
 
-    if 'diskSpace' not in config:
-        config.put('diskSpace', BNDL_DEFAULT_DISK_SPACE)
 
-    if 'memory' not in config:
-        config.put('memory', BNDL_DEFAULT_MEMORY)
+def load_bundle_args_into_conf(config, args, with_defaults):
+    # this is unrolled because it's actually pretty complicated to get the order
+    # correct given that some attributes need special handling and defaults
 
-    if 'nrOfCpus' not in config:
-        config.put('nrOfCpus', BNDL_DEFAULT_NR_OF_CPUS)
+    args_compatibility_version = getattr(args, 'compatibility_version', None)
+    args_disk_space = getattr(args, 'disk_space', None)
+    args_memory = getattr(args, 'memory', None)
+    args_name = getattr(args, 'name', None)
+    args_nr_of_cpus = getattr(args, 'nr_of_cpus', None)
+    args_roles = getattr(args, 'roles', None)
+    args_system = getattr(args, 'system', None)
+    args_system_version = getattr(args, 'system_version', None)
+    args_version = getattr(args, 'version', None)
 
-    if 'version' not in config:
-        config.put('version', BNDL_DEFAULT_VERSION)
+    if hasattr(args, 'annotations') and len(args.annotations) > 0:
+        annotations_tree = ConfigTree()
 
-    if 'system' not in config:
-        config.put('system', config.get('name'))
+        for annotation in args.annotations:
+            if '=' not in annotation:
+                raise ValueError('Invalid annotation format. Specify as name=value')
```

Comment:
I'd suggest displaying the offending `annotation` in the error message, i.e.

```
raise ValueError('Invalid annotation format for {}. Specify as name=value'.format(annotation))
```

Also perhaps we can collect all offending annotations and display them altogether, i.e.

```
invalid_annotations = []
for annotation in args.annotations:
  if '=' not in annotation:
    invalid_annotation.append(annotation)

if invalid_annotations:
  display_invalid_annotations = ', '.join(invalid_annotations)  
  raise ValueError('Invalid annotation format for {}. Specify as name=value'.format(display_invalid_annotations))

...
```

Otherwise if there are multiple invalid annotation specified, the user will be prompted each of these error one at a time.